src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Cohost/RazorCohostCapabilitiesProvider.cs

```diff
@@ -12,15 +13,23 @@
 namespace Microsoft.VisualStudio.LanguageServerClient.Razor.Cohost;
 
 [Export(typeof(IRazorCohostCapabilitiesProvider)), System.Composition.Shared]
+[Export(typeof(IClientCapabilitiesService))]
 [method: ImportingConstructor]
-internal class RazorCohostCapabilitiesProvider([ImportMany(typeof(ICapabilitiesProvider))] IEnumerable<Lazy<ICapabilitiesProvider>> cohostCapabilitiesProviders) : IRazorCohostCapabilitiesProvider
+internal class RazorCohostCapabilitiesProvider([ImportMany(typeof(ICapabilitiesProvider))] IEnumerable<Lazy<ICapabilitiesProvider>> cohostCapabilitiesProviders)
+    : IRazorCohostCapabilitiesProvider, IClientCapabilitiesService
 {
     private readonly IEnumerable<Lazy<ICapabilitiesProvider>> _cohostCapabilitiesProviders = cohostCapabilitiesProviders;
 
+    private VSInternalClientCapabilities? _clientCapabilities;
+
+    public VSInternalClientCapabilities ClientCapabilities => _clientCapabilities.AssumeNotNull();
```

Comment:
>ClientCapabilities [](http://example.com/codeflow?start=40&length=18)

I'm trying to understand the model here. Can we have multiple clients connected and/or one client at at time but more than one to the same server? This seems to return "last connected client capabilities", is that right - or am I misunderstanding how this works? Maybe better to discuss offline. 

My hesitation here is that the value we return here depends on the value we set in the last GetCapabilities call, and it will throw if we try to get it before that call, so there are sequencing constraints here. Basically GetCapabilities is more like "InitializeClientCapabilities".

Reply:
I'll put my 2c in, since this may also help drive discussion (and let David tell me if I'm wrong). 

We're using `GetCapabilities` to fill in our `VSInternalServerCapabilities`. We just happen to store the client capabilities passed to use for that during that time. This should be part of LSP initialization, so we assume that it's called before any of our endpoints are hooked up (because we haven't advertised any capabilities yet). That's why we can `AssumeNotNull` here.

Reply:
You're all correct, including your observation Alex that its a little weird using GetCapabilities to fill this in. The non-cohost impl is a much better model, but right now with cohosting there is no way to do it that way, and it requires a Roslyn change and version bump. We'll get there eventually though, as we'll need to do something similar for dynamic registration.

So the model here is not that there are multiple clients at the same time, just that each client that could use a language server has different capabilities. The way LSP works is the first thing after a server is started (the client owns the lifetime of the server) is it sends an `initialize` message, where it tells the server what it is capable of. The server responds to that with the server capabilities, and from then on messages just flow freely. So if we, as the server, want to be able to reason about the clients capabilities later on we have to capture them from the `initialize` message. The `IInitializeManager<T, T2>` does that in non-cohost world, but in cohosting, while that interface does exist and work in the same way, its all in Roslyn and we can't access it yet.

`GetCapabilities` is the method by which we, as the server, look through all of our endpoints to constuct the _server_ capabiltiies. It just so happens though, that it's a convenient spot where we get a callback inside the `initialize` handling, that passes the client capabilities around, so I can piggy-back on it in this slightly hacky way.

Also, yes this is just the "most recent" capabilities, depending on service lifetimes, and yes things throw if anything tries to access capabilities before the `initialize` message (which is why there is a `Lazy` in the semantic tokens legend).